src/geom/detail/frameSetUtils.cc

```diff
@@ -0,0 +1,246 @@
+        throw LSST_EXCEPT(lsst::pex::exceptions::InvalidParameterError, os.str());
+    }
+    if (strip) {
+        auto const finalNames = setFromVector(channel.getCardNames());
+
+        // FITS keywords that FitsChan stripped
+        std::set<std::string> namesChannelStripped;
+        std::set_difference(initialNames.begin(), initialNames.end(), finalNames.begin(), finalNames.end(),
+                            std::inserter(namesChannelStripped, namesChannelStripped.begin()));
+
+        // FITS keywords that FitsChan may strip that we want to keep in `metadata`
+        // TODO remove TIMESYS from this list once we update AST 2017-04-24 or newer
+        std::set<std::string> const namesToKeep = {"DATE-OBS", "TIMESYS"};
+
+        std::set<std::string> namesToStrip;  // names to strip from metadata
+        std::set_difference(namesChannelStripped.begin(), namesChannelStripped.end(), namesToKeep.begin(),
+                            namesToKeep.end(), std::inserter(namesToStrip, namesToStrip.begin()));
+        for (auto const& name : namesToStrip) {
+            metadata.remove(name);
+        }
+    }
+    return frameSet;
+}
+
+std::shared_ptr<ast::FrameSet> readLsstSkyWcs(daf::base::PropertyList& metadata, bool strip) {
+    std::vector<double> crvalRad = {degToRad(metadata.getAsDouble("CRVAL1")),
+                                    degToRad(metadata.getAsDouble("CRVAL2"))};
+    auto frameSet = readFitsWcs(metadata, strip);
+
+    // Standardize the Sky frame to ICRS with axes order RA, Dec,
+    // but inherit what we can from the original sky frame (e.g. epoch)
+
+    // Create the desired sky frame, to use as a template
+    auto stdSkyFrame = ast::SkyFrame("System=ICRS");
+
+    // Locate a Frame in the target FrameSet that looks like the template
+    // and hence can be used as the original sky frame.
+    // We ignore the frame set returned by findFrame because that goes from pixels to sky,
+    // and using it would add an unwanted extra branch to our WCS; instead, later on,
+    // we compute a mapping from the old sky frame to the new sky frame and add that.
+    if (!frameSet->findFrame(stdSkyFrame)) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "Could not find a SkyFrame");
+    }
+    auto initialSkyIndex = frameSet->getCurrent();
+
+    // Compute a frame set that maps from the original sky frame to our desired sky frame;
+    // this is what we will insert into the frame set.
+    // (Temporarily set the base frame to the sky frame, because findFrame
+    // produces a mapping from base to the found frame).
+    auto initialBaseIndex = frameSet->getBase();
+    frameSet->setBase(initialSkyIndex);
+    auto stdSkyFrameSet = frameSet->findFrame(stdSkyFrame);
+    if (!stdSkyFrameSet) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError,
+                          "Bug: found a SkyFrame the first time, but not the second time");
```

Comment:
I feel I've done what I can in the code comments before both blocks of code that call findFrame. I agree it's weird.